enhancements/update/update-status-api.md

```diff
@@ -0,0 +1,410 @@
+
+### Risks and Mitigations
+
+The proposal to deliver the API and a controller that both manages the API and monitors the cluster (producing insights) before achieving consensus on the eventual modular system architecture risks that the existing API will not accommodate the future architecture well. We are making a tradeoff to deliver the API early to start providing value, which also allows us to learn about how and if such an API is really consumed. Early delivery directly addresses the risk of investing effort into building a much larger system that may not address real user needs.
+
+#### API Size and Footprint
+
+The instances of `UpdateStatus` API, as proposed in https://github.com/openshift/api/pull/2012, are expected to be quite large. A single `UpdateStatus` resource status is a tree of the following shape, with update insight structures as leaves:
+
+```
+status:
+  # two nodes on this level
+  controPlane:
+    ...
+    informers:
+    # one node per informer reporting control plane insights, typically ~2-5
+    - name: informer-1
+      insights:
+      # one leaf per insight
+  workerPools:
+  # one node per pool
+  - name: mcp-1
+    informers:
+    # one node per informer reporting pool insights, typically ~2-5
+    - name: informer-2
+      insights
+      # one leaf per insight
+```
+
+Health insights are not expected to be created in large numbers (like alerts, they are not helpful when too many are reported). A healthy update should have zero. The aspiration for the system is to produce precisely one health insight for a problematic update: the one that clearly describes the problem affecting the update. We are unlikely to achieve this ideal, but a typical problematic update should produce no more than ~10 health insights.
+
+However, there is a single Status Insight for each resource involved in the update, so we can expect `1 + # of ClusterOperators + # of MachineConfigPools + # of Nodes` status insights. That means that for big clusters, there will always be (lower) hundreds of insights, primarily because of Nodes.
+
+The Status Insights are not expected to be large data structures, but some contain Kubernetes conditions, so they cannot be considered entirely trivial, and there can be a high number of them. The Health Insights are larger because they can carry a longer-form description of the issue, expected to be around a paragraph of text.
+
+For an API this large, we need to consider the possibility of hitting Kubernetes API resource limits. Even with too many long insights, hitting the 1.5MiB limit does not seem plausible under regular operation. We can protect against the risk by setting appropriate bounds and limits on API fields (especially lists and the long-form description field of health insights), and these limits will be part of the contract for insight producers.
```

Comment:
I would be interested to see some numbers to support this, have any actual calculations been done?

Reply:
I did some vague ones but I do not have the numbers anymore. I'll get them again, but maybe I'll just make a test that produces a worst-case specimen and measures its size.